### PR TITLE
Fix undefined variable and write unit tests for OC_Principal connector

### LINKED ISSUE #12681: server to server sharing from oc8 to oc7 - error: "undefined variable: user"
Steps

Setup OC1 at Version 7.0.2 and OC2 at Daily-Master. 
Add a Folder in OC2 and share it via link,  with password and editable (rw)  for users.
Open the link,  add to owncloud OC1. Add some files into the shared folder from within OC2 and from OC1.

Expected result

File can be downloaded from OC2 via OC1
Files can be added through OC1

Actual result

Both do not work. One can see all files created in OC2, but downloads fail given the error "undefined variable: user"  in sabre dav.  Uploading files to OC2
 over OC1 never finishes uploading. 

Versions

OC1 was 7.0.2
OC2 was pre8

Explanation

I have shared a folder from an OC 8 (yesterdays build) to an OC 7.0.2 instance (web hosted). I can connect those folders and browse it from the OC7 instance, but if i want to download or open a file, i get an error. Here´s the Trace:

```
{"reqId":"6d759195bb3a5079982829f70ef9eff0","remoteAddr":"2001:4ba0:xxxx:139::250:xxxx","app":"webdav","message":"Exception: {\"Message\":\"\",\"Code\":0,\"Trace\":\"#0 \\/var\\/www\\/owncloud\\/3rdparty\\/sabre\\/dav\\/lib\\/Sabre\\/DAV\\/Server.php(530): OC_Connector_Sabre_File->get()\n#1 \\/var\\/www\\/owncloud\\/lib\\/private\\/connector\\/sabre\\/server.php(89): Sabre\\DAV\\Server->httpGet('etrij.oct2010.0...')\n#2 [internal function]: OC_Connector_Sabre_Server->httpGet('etrij.oct2010.0...')\n#3 \\/var\\/www\\/owncloud\\/3rdparty\\/sabre\\/dav\\/lib\\/Sabre\\/DAV\\/Server.php(474): call_user_func(Array, 'etrij.oct2010.0...')\n#4 \\/var\\/www\\/owncloud\\/3rdparty\\/sabre\\/dav\\/lib\\/Sabre\\/DAV\\/Server.php(214): Sabre\\DAV\\Server->invokeMethod('GET', 'etrij.oct2010.0...')\n#5 \\/var\\/www\\/owncloud\\/apps\\/files_sharing\\/publicwebdav.php(59): Sabre\\DAV\\Server->exec()\n#6 \\/var\\/www\\/owncloud\\/public.php(48): require_once('\\/var\\/www\\/ownclo...')\n#7 {main}\",\"File\":\"\\/var\\/www\\/owncloud\\/lib\\/private\\/connector\\/sabre\\/file.php\",\"Line\":165}","level":4,"time":"December 03, 2014 23:06:56"}
```
```
{"reqId":"2fa6dcbeeff48ee3ce528758a34bf001","remoteAddr":"2a02:2450:xxxx:1fd:xxxx:xxxx:xxxx:3d02","app":"PHP","message":"Undefined variable: user at \/var\/www\/owncloud\/lib\/private\/connector\/sabre\/principal.php#56","level":3,"time":"December 03, 2014 23:07:01"}
```

### PR DESCRIPTION
**Pull request by MorrisJobke**
Superseeds #12960 - Rebased - Unit tests pass locally. (I also did this to have a working Jenkins run for this)

fixes #12681 properly.

Improves the PHPDoc and adds tests to this code path.

@karlitschek I vote for merging this, as it fixes the bug #12681 properly and introduces tests for this code path. The decision about merge this for ownCloud 8 is upon you. Most changes are unit tests and PHPDoc updates

cc @PVince81 @DeepDiver1975 @icewind1991 Please review again.

### CODE REVIEW

**Comment by MorrisJobke** [comment]
@LukasReschke Please also have a look

**Comment by MorrisJobke** [comment]
@derkostka Can I ask you to test this too? Thanks

**Comment by LukasReschke** [comment]
THX for rebasing - not sure if I can judge about code since it is from me but rebasing seems fine :+1:

**Comment by MorrisJobke** [comment]
@LukasReschke I just want to be sure to have done it in the correct way ;)

**Comment by karlitschek** [comment]
@morrisjobke :+1:

**Comment by derkostka** [comment]
@MorrisJobke  :+1: I tested it based on OC8-Alpha1. All well !

**Comment by derkostka** [comment]
@MorrisJobke @LukasReschke calendar app seems to be affected by this change:

{"reqId":"44aa732da52b771c9db71da8c82c22f8","remoteAddr":"2a02:2450:102d:1fd:fce6:d5b4:70c1:3484","app":"PHP","message":"Class 'OC_Connector_Sabre_Principal' not found at \/var\/www\/owncloud\/apps\/calendar\/appinfo\/remote.php#23","level":3,"time":"January 05, 2015 19:50:43"}
occurs when syncing calendar ...

**Comment by LukasReschke** [comment]
Well - the fixes are already linked in the original PR ;-)

File: lib/private/connector/sabre/principal.php [hunk 1]
```diff
@@ -2,12 +2,34 @@
...
  * Copyright (c) 2011 Jakob Sack mail@jakobsack.de
  * Copyright (c) 2012 Bart Visscher <bartv@thisnet.nl>
+ * Copyright (c) 2014 Lukas Reschke lukas@owncloud.com
+ *
  * This file is licensed under the Affero General Public License version 3 or
  * later.
  * See the COPYING-README file.
  */
 
-class OC_Connector_Sabre_Principal implements \Sabre\DAVACL\PrincipalBackend\BackendInterface {
```
Comment (derkostka) [inline lib/private/connector/sabre/principal.php]: class is referenced by calendar\/appinfo\/remote.php#23

**Comment by LukasReschke** [comment]
From the original PR:

@schiesbn The principal connector has now 100% test coverage, but it now requires DI thus https://github.com/owncloud/calendar/pull/655 and https://github.com/owncloud/contacts/pull/749 needs to get merged as well.

Please review and test.

**Comment by derkostka** [comment]
:+1:  from me, calendar sync is functional with both PRs

**Comment by LukasReschke** [comment]
No worries :-)

**Pull request merged**